### PR TITLE
Optimize StreamInput#readString

### PR DESCRIPTION
**Pull request by original-brownbear**
* Resolve TODO in `readString` by moving to reading chunks of `byte[]` instead of going byte by byte
* Motivated by `readString` showing up as a significant user of CPU time on the IO thread in Rally PMC benchmark
* Benchmarking this:
  * Could not reproduce a slowdown in the potential worst case (one or two non-ascii chars) since in this case the cost of creating the string itself exceeds the read times anyway
  * Speedup for 50%+ for reading 200 char ascii strings from `ByteBuf` or pages bytes backed streams
  * Longer strings obviously get bigger speedups
  * More ascii chars -> more speedup

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -419,34 +420,119 @@ public String readString() throws IOException {
...
+                } else {
+                    toRead = bufferFree;
+                }
+            } else {
+                toRead = minRemainingBytes;
+            }
+            readBytes(byteBuffer, limitByteArray, toRead);
+            limitByteArray += toRead;
+            int bufferedBytesRemaining;
+            // As long as we at least have three bytes buffered we don't need to do any bounds checking when getting the next byte since we
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: next byte -> next character?

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 2]
```diff
@@ -419,34 +420,119 @@ public String readString() throws IOException {
...
-                    buffer[i] = ((char) ((c & 0x1F) << 6 | readByte() & 0x3F));
-                    break;
-                case 14:
-                    buffer[i] = ((char) ((c & 0x0F) << 12 | (readByte() & 0x3F) << 6 | (readByte() & 0x3F) << 0));
-                    break;
-                default:
-                    throw new IOException("Invalid string; unexpected character: " + c + " hex: " + Integer.toHexString(c));
+        int charsProduced = 0;
+        int posByteArray = 0;
+        int limitByteArray = 0;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: I think the name `usedByteBuffer` would make more sense, I had to look at some uses to understand this was the end of the used portion of the buffer

**Review by rjernst — COMMENTED** [review]
This looks like a nice perf improvement! I have two major thoughts (I haven't looked closely yet, but wanted to get these thoughts out there for discussion):
* This is quite a complicated change, so I think we should have tests of StreamInput specifically covering all the corner cases?
* I wonder if we could simplify all of this by changing the wire format for strings to store the byte count instead of char count. Since we take in a String for writing, we already have the entire thing in memory, so converting to a byte buffer doesn't seem like to much. We already use a scratch buffer, which will be big enough for the vast majority of the strings we have, but if it is not, we can size appropriately? Alternatively, we could write out strings in chunks, where each buffer of output is lead by the length of that buffer.

**Comment by original-brownbear** [comment]
@rjernst thanks for taking a look!

> This is quite a complicated change, so I think we should have tests of StreamInput specifically covering all the corner cases?

I looked into this, but in the end I decided I couldn't find a corner case not covered by `org.elasticsearch.common.io.stream.BytesStreamsTests#testWriteRandomStrings`. That test has up to twice the length of our thread-local char buffer and by also having 2048 chars max we cover all the possible scenarios of reusing the `byte[]` I think. As far as I can see we def. have 100% of lines in the new code covered and I can't really see a logical scenario not covered either but I could be missing some obvious one?

> I wonder if we could simplify all of this by changing the wire format for strings to store the byte count instead of char count. Since we take in a String for writing, we already have the entire thing in memory, so converting to a byte buffer doesn't seem like to much. We already use a scratch buffer, which will be big enough for the vast majority of the strings we have, but if it is not, we can size appropriately? Alternatively, we could write out strings in chunks, where each buffer of output is lead by the length of that buffer.

Yea I had some exchange with Simon on this before as well. Initially, that ended up with maybe going for the chunk by chunk approach. I'm still not convinced that gets us much though. The problem is this (maybe you have solutions though?):

* If we want to encode the full size of the string and write it in one go, we have a lot of extra allocations for large strings. The whole point of optimizing this though, is making large strings faster which show much bigger performance gains so I'm not that big of a fan of doing it all in one go.
* Chunked encoding seems better, but it's still tricky to encode the length. We can't put the length of the full string into the first chunk since we don't know it yet (unless we want to go through all the string's chars twice). So my initial though was to just take 2048 byte chunks and use the first two bytes as a header. Encode whether or not it's the last chunk in the string via the first bit and use the remaining 13 bits to encode the length of the chunk (since that will vary a little depending on the kind of char for full chunks and we have the final chunk as well that could be smaller). But then it's tricky to pre-allocate the correct `char[]` for long strings and we get a lot of allocations and copying for those in `StreamInput`. It's hard for me to judge the tradeoff here. For shorter strings just encoding the byte length is a significant win compared to current `master` but compared to the solution here the gain isn't that large (and for ascii strings it's non-existent pretty much). For longer strings I'm not sure we'll gain much in the real world unless we read a bunch of long string back to back from the same `StreamInput` so we can reuse the large `char[]` we're reading into. If we just read one large string we're more efficient in reading the bytes, but we pay for that with resizing the `char[]` buffer that we now don't know the size of. 

... that's the reason I pretty much gave up on chunking in theory. Also, I couldn't really figure out a good benchmark here to make a judgement. This solution just wins over the existing in all cases pretty much or is at least equal, but for chunking it's easy to create a case where this one is faster (as well as cases where it's the other way around ...).

**Comment by rjernst** [comment]
> unless we want to go through all the string's chars twice

Maybe this wouldn't be so bad? I'd at least be curious what the perf is in this case for comparison?

**Comment by original-brownbear** [comment]
> Maybe this wouldn't be so bad? I'd at least be curious what the perf is in this case for comparison?

It depends: 
One important thing to note about the real world here is that we're currently using `vInt` to encode the string length in chars.
In the PMC benchmark ~36% of the runtime in `readString` is actually spent reading the `vInt` according to YourKit tracing. A little over ~50% is spend on reading byte arrays. For `writeString` it's evenly split on the IO side with ~45% each going into IO.

So, I did some quick and dirty testing only here so numbers aren't super accurate (and are only tested with Netty `ByteBuf` backed inputs ... which are the only thing that matters in the real world right now though) but:
If I encode the length as `vInt` still I couldn't make out much difference in throughput. Reads maybe get a little faster in some cases (long strings with few ascii chars) and stay the same for ascii chars compared to the solution here. Writes get a little slower, but also not by much.

=> I'm between not sure and don't care if we keep prefixing with `vInt` for the byte count.

*But*, if we were not to use `vInt` but just `int` for the length prefix (like we do for `Text`!) the situation changes quite a bit:
Now we get a significant speedup at the cost of 1-3 added bytes per string. Half, the cost of each Netty `ByteBuf` operation is made up of the constant cost of (since we mostly read from unpooled or a single pool chunk buffer) `io.netty.buffer.AbstractByteBuf.checkReadableBytes(int)` => if we encode the length as `int` we call that a lot less (only once per int) and then mostly only make a single read into a `byte[]`.
Also, you can do a smarter write side here as well for strings that fit into one buffer if you use an `int` prefix. You can simply start putting the chars into the `byte[]` buffer starting at index 4, then populate the int at the beginning when done with that -> for strings that fit into the buffer there's no double processing and for those that don't you at least don't have to double process part of the string.

=> maybe the bandwidth hit of not using `vInt` and using `int` for the prefix is worth the speedup here if we actually want to improve throughput more? If not, why do we `int` prefix `Text`?

**Comment by rjernst** [comment]
> maybe the bandwidth hit of not using vInt and using int for the prefix is worth the speedup 

+1 to keep exploring this! Probably worth a discussion with the team to find if anyone else knows of reasons we should favor vint for string length.

**Comment by original-brownbear** [comment]
@rjernst so I had a little time to look into a different wire format now, using an `int` for the number of `bytes` as a prefix to get better bulk reading.

The results of that weren't so stellar:
* Read performance across the board doesn't get a huge boost if any:
  * Since we don't have the char count now (and can't build our the `String` directly from `byte[]` as a result of our encoding) we still have all the same steps we had before on the read side, and save a small number of reads only (none for ascii strings and then even for strings with non-ascii chars only it's only a factor of less than 10x) 
  * We can't size the `CharsRef` as optimally as we could before for long strings (because now we don't know the exact char count)

On the other hand, write performance takes a measurable hit for longer strings:
 * Obviously, even if done optimally by not double processing a full buffer size to get the length, all `Strings` that need more than one write buffer size cost a bit extra

-> going for a full `int` prefix and fewer by putting the bytes in the beginning of the write looked better in the profiler than in the real world. In the end, this change gets a big gain out of reducing the number of reads by a large factor (for ascii strings it's almost the number of chars) but additional improvements don't add much since the overhead of the other operations comes into play.

-> I'd rather go with this than change the format. Changing the format just adds risk and I don't see how the byte count prefix really gets us much in practice (for reading a single long string from a stream it rather seems to hurt us because we can't efficiently allocate the `char[]` buffer).

**Review by rjernst — APPROVED** [review]
LGTM. I left a few comments.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/docs

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/docs

**Comment by original-brownbear** [comment]
Thanks @rjernst, applied changes as discussed -> merging :)

**Pull request merged**